File: stacks/python/base/code/test_integration.tpl.py
"""
File: test_integration.tpl.py
Purpose: Integration test patterns and examples
Generated for: {{PROJECT_NAME}}
"""

import pytest
from httpx import AsyncClient


# ============================================================================
# API Integration Tests
# ============================================================================

class TestAuthenticationFlow:
    """Integration tests for authentication flow."""
    
    @pytest.mark.asyncio
    async def test_register_login_flow(self, client: AsyncClient):
        """Test complete registration and login flow."""
        # Register
        register_data = {
            "email": "newuser@test.com",
            "username": "newuser",
            "password": "SecurePassword123!",
            "full_name": "New User",
        }
        response = await client.post("/api/v1/auth/register", json=register_data)
        assert response.status_code == 201
        data = response.json()
        assert "access_token" in data
        user_id = data["user"]["id"]
        
        # Login with credentials
        login_data = {
            "email": register_data["email"],
            "password": register_data["password"],
        }
        response = await client.post("/api/v1/auth/login", json=login_data)
        assert response.status_code == 200
        data = response.json()
        assert "access_token" in data
        token = data["access_token"]
        
        # Access protected resource
        headers = {"Authorization": f"Bearer {token}"}
        response = await client.get("/api/v1/users/me", headers=headers)
        assert response.status_code == 200
        assert response.json()["id"] == user_id
    
    @pytest.mark.asyncio
    async def test_password_reset_flow(self, client: AsyncClient, test_user, mock_email_service):
        """Test password reset flow."""
        # Request password reset
        response = await client.post(
            "/api/v1/auth/password-reset/request",
            json={"email": test_user.email}
        )
        assert response.status_code == 200
        
        # Verify email was sent (check mock)
        mock_email_service.send_template.assert_called_once()
        
        # Get reset token from mock call
        call_args = mock_email_service.send_template.call_args
        reset_token = call_args.kwargs.get("data", {}).get("reset_token")
        
        # Reset password
        new_password = "NewSecurePassword456!"
        response = await client.post(
            "/api/v1/auth/password-reset/confirm",
            json={"token": reset_token, "new_password": new_password}
        )
        assert response.status_code == 200
        
        # Login with new password
        response = await client.post(
            "/api/v1/auth/login",
            json={"email": test_user.email, "password": new_password}
        )
        assert response.status_code == 200


class TestCRUDOperations:
    """Integration tests for CRUD operations."""
    
    @pytest.mark.asyncio
    async def test_create_read_update_delete_post(
        self, authenticated_client: AsyncClient, test_user
    ):
        """Test complete CRUD flow for posts."""
        # Create
        post_data = {
            "title": "Test Post",
            "content": "This is the content of the test post.",
            "status": "draft",
        }
        response = await authenticated_client.post("/api/v1/posts", json=post_data)
        assert response.status_code == 201
        post = response.json()
        post_id = post["id"]
        assert post["title"] == post_data["title"]
        assert post["author_id"] == test_user.id
        
        # Read
        response = await authenticated_client.get(f"/api/v1/posts/{post_id}")
        assert response.status_code == 200
        assert response.json()["id"] == post_id
        
        # Update
        update_data = {"title": "Updated Title", "status": "published"}
        response = await authenticated_client.patch(
            f"/api/v1/posts/{post_id}", json=update_data
        )
        assert response.status_code == 200
        assert response.json()["title"] == "Updated Title"
        assert response.json()["status"] == "published"
        
        # List
        response = await authenticated_client.get("/api/v1/posts")
        assert response.status_code == 200
        posts = response.json()["data"]
        assert any(p["id"] == post_id for p in posts)
        
        # Delete
        response = await authenticated_client.delete(f"/api/v1/posts/{post_id}")
        assert response.status_code == 204
        
        # Verify deleted
        response = await authenticated_client.get(f"/api/v1/posts/{post_id}")
        assert response.status_code == 404
    
    @pytest.mark.asyncio
    async def test_pagination(self, authenticated_client: AsyncClient, post_factory, test_user):
        """Test pagination of list endpoints."""
        # Create multiple posts
        for _ in range(25):
            await post_factory.create(author=test_user)
        
        # First page
        response = await authenticated_client.get("/api/v1/posts?page=1&per_page=10")
        assert response.status_code == 200
        data = response.json()
        assert len(data["data"]) == 10
        assert data["pagination"]["page"] == 1
        assert data["pagination"]["total"] >= 25
        
        # Second page
        response = await authenticated_client.get("/api/v1/posts?page=2&per_page=10")
        assert response.status_code == 200
        data = response.json()
        assert len(data["data"]) == 10
        assert data["pagination"]["page"] == 2


class TestConcurrency:
    """Integration tests for concurrent operations."""
    
    @pytest.mark.asyncio
    async def test_concurrent_updates(self, authenticated_client: AsyncClient, post_factory, test_user):
        """Test handling of concurrent updates."""
        import asyncio
        
        post = await post_factory.create(author=test_user, view_count=0)
        
        # Simulate concurrent view increments
        async def increment_view():
            await authenticated_client.post(f"/api/v1/posts/{post.id}/view")
        
        tasks = [increment_view() for _ in range(10)]
        await asyncio.gather(*tasks)
        
        # Verify final count
        response = await authenticated_client.get(f"/api/v1/posts/{post.id}")
        assert response.json()["view_count"] == 10
    
    @pytest.mark.asyncio
    async def test_rate_limiting(self, client: AsyncClient):
        """Test rate limiting."""
        import asyncio
        
        async def make_request():
            return await client.get("/api/v1/health")
        
        # Make many requests quickly
        tasks = [make_request() for _ in range(100)]
        responses = await asyncio.gather(*tasks)
        
        # Some should be rate limited
        status_codes = [r.status_code for r in responses]
        assert 429 in status_codes, "Rate limiting should be triggered"


class TestWebhooks:
    """Integration tests for webhook handling."""
    
    @pytest.mark.asyncio
    async def test_stripe_webhook(self, client: AsyncClient, mock_payment_service):
        """Test Stripe webhook handling."""
        # Simulate Stripe webhook payload
        payload = {
            "type": "customer.subscription.updated",
            "data": {
                "object": {
                    "id": "sub_123",
                    "customer": "cus_456",
                    "status": "active",
                }
            }
        }
        
        # Send webhook with signature
        signature = "test_signature"  # In real test, compute proper signature
        headers = {"Stripe-Signature": signature}
        
        response = await client.post(
            "/api/v1/webhooks/stripe",
            json=payload,
            headers=headers
        )
        assert response.status_code == 200


class TestFileUpload:
    """Integration tests for file uploads."""
    
    @pytest.mark.asyncio
    async def test_upload_and_download(self, authenticated_client: AsyncClient):
        """Test file upload and download."""
        # Upload
        files = {"file": ("test.txt", b"Hello, World!", "text/plain")}
        response = await authenticated_client.post("/api/v1/files/upload", files=files)
        assert response.status_code == 201
        file_id = response.json()["id"]
        
        # Download
        response = await authenticated_client.get(f"/api/v1/files/{file_id}/download")
        assert response.status_code == 200
        assert response.content == b"Hello, World!"
    
    @pytest.mark.asyncio
    async def test_upload_size_limit(self, authenticated_client: AsyncClient):
        """Test file size limit enforcement."""
        # Create large file (assuming 10MB limit)
        large_content = b"x" * (11 * 1024 * 1024)  # 11 MB
        files = {"file": ("large.bin", large_content, "application/octet-stream")}
        
        response = await authenticated_client.post("/api/v1/files/upload", files=files)
        assert response.status_code == 413  # Payload Too Large


# ============================================================================
# Database Integration Tests
# ============================================================================

class TestDatabaseIntegrity:
    """Tests for database integrity and constraints."""
    
    @pytest.mark.asyncio
    async def test_unique_email_constraint(self, db_session, user_factory):
        """Test unique email constraint."""
        user1 = await user_factory.create(email="duplicate@test.com")
        
        with pytest.raises(Exception) as exc_info:
            await user_factory.create(email="duplicate@test.com")
        
        assert "unique" in str(exc_info.value).lower()
    
    @pytest.mark.asyncio
    async def test_cascade_delete(self, db_session, user_factory, post_factory):
        """Test cascade delete behavior."""
        from sqlalchemy import select
        
        user = await user_factory.create()
        post1 = await post_factory.create(author=user)
        post2 = await post_factory.create(author=user)
        
        # Delete user
        await db_session.delete(user)
        await db_session.flush()
        
        # Posts should be deleted
        from models import Post
        result = await db_session.execute(
            select(Post).filter_by(author_id=user.id)
        )
        posts = result.scalars().all()
        assert len(posts) == 0
    
    @pytest.mark.asyncio
    async def test_soft_delete(self, db_session, post_factory, test_user):
        """Test soft delete functionality."""
        from sqlalchemy import select
        
        post = await post_factory.create(author=test_user)
        post_id = post.id
        
        # Soft delete
        post.soft_delete()
        await db_session.flush()
        
        # Should not appear in normal queries
        from models import Post
        result = await db_session.execute(
            select(Post).filter_by(id=post_id, is_deleted=False)
        )
        assert result.scalar_one_or_none() is None
        
        # Should appear when including deleted
        result = await db_session.execute(
            select(Post).filter_by(id=post_id)
        )
        deleted_post = result.scalar_one()
        assert deleted_post.is_deleted is True

File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/node/base/code/test_integration.tpl.js
/*
File: test_integration.tpl.js
Purpose: Integration test patterns and examples
Generated for: {{PROJECT_NAME}}
*/

const {
    getPrisma,
    setupDatabase,
    teardownDatabase,
    cleanDatabase,
    createUser,
    createPost,
    createAuthToken,
    createTestClient,
    faker,
} = require('./test_fixtures');

const { assertResponse, assertExistsInDb, waitFor } = require('./test_helpers');

// Import your app
// const app = require('../src/app');

// ============================================================================
// Setup
// ============================================================================

describe('Integration Tests', () => {
    let prisma;
    let client;

    beforeAll(async () => {
        prisma = await setupDatabase();
        // client = createTestClient(app);
    });

    afterAll(async () => {
        await teardownDatabase();
    });

    beforeEach(async () => {
        await cleanDatabase();
    });

    // ============================================================================
    // Authentication Flow Tests
    // ============================================================================

    describe('Authentication Flow', () => {
        describe('POST /api/auth/register', () => {
            it('should register a new user', async () => {
                const userData = {
                    email: 'newuser@test.com',
                    username: 'newuser',
                    password: 'SecurePassword123!',
                    fullName: 'New User',
                };

                const response = await client.post('/api/auth/register', userData);

                assertResponse(response)
                    .created()
                    .hasKey('user', 'accessToken')
                    .dataContains({ email: userData.email });

                await assertExistsInDb(prisma, 'user', { email: userData.email });
            });

            it('should reject duplicate email', async () => {
                const existingUser = await createUser(prisma, { email: 'existing@test.com' });

                const response = await client.post('/api/auth/register', {
                    email: 'existing@test.com',
                    username: 'newuser',
                    password: 'Password123!',
                });

                assertResponse(response)
                    .badRequest()
                    .hasError('EMAIL_EXISTS');
            });

            it('should validate password requirements', async () => {
                const response = await client.post('/api/auth/register', {
                    email: 'test@test.com',
                    username: 'testuser',
                    password: '123', // Too weak
                });

                assertResponse(response)
                    .unprocessable()
                    .hasError('VALIDATION_ERROR');
            });
        });

        describe('POST /api/auth/login', () => {
            it('should login with valid credentials', async () => {
                const password = 'TestPassword123!';
                const user = await createUser(prisma, {
                    email: 'test@test.com',
                    // In real tests, hash the password properly
                });

                const response = await client.post('/api/auth/login', {
                    email: user.email,
                    password,
                });

                assertResponse(response)
                    .ok()
                    .hasKey('accessToken', 'refreshToken', 'user');
            });

            it('should reject invalid credentials', async () => {
                const user = await createUser(prisma);

                const response = await client.post('/api/auth/login', {
                    email: user.email,
                    password: 'wrongpassword',
                });

                assertResponse(response)
                    .unauthorized()
                    .hasError('INVALID_CREDENTIALS');
            });

            it('should reject inactive user', async () => {
                const user = await createUser(prisma, { isActive: false });

                const response = await client.post('/api/auth/login', {
                    email: user.email,
                    password: 'TestPassword123!',
                });

                assertResponse(response)
                    .forbidden()
                    .hasError('ACCOUNT_DISABLED');
            });
        });

        describe('Protected Routes', () => {
            it('should access protected route with valid token', async () => {
                const user = await createUser(prisma);
                const token = createAuthToken(user.id);

                const response = await client.authGet('/api/users/me', token);

                assertResponse(response)
                    .ok()
                    .dataContains({ id: user.id, email: user.email });
            });

            it('should reject expired token', async () => {
                const user = await createUser(prisma);
                const token = createAuthToken(user.id, { expiresIn: '-1h' });

                const response = await client.authGet('/api/users/me', token);

                assertResponse(response).unauthorized();
            });

            it('should reject missing token', async () => {
                const response = await client.get('/api/users/me');

                assertResponse(response).unauthorized();
            });
        });
    });

    // ============================================================================
    // CRUD Operations Tests
    // ============================================================================

    describe('CRUD Operations', () => {
        let user;
        let token;

        beforeEach(async () => {
            user = await createUser(prisma);
            token = createAuthToken(user.id);
        });

        describe('Posts', () => {
            describe('POST /api/posts', () => {
                it('should create a new post', async () => {
                    const postData = {
                        title: 'Test Post',
                        content: 'This is the content.',
                        status: 'draft',
                    };

                    const response = await client.authPost('/api/posts', postData, token);

                    assertResponse(response)
                        .created()
                        .dataContains({
                            title: postData.title,
                            authorId: user.id,
                        });

                    const post = response.body.data;
                    await assertExistsInDb(prisma, 'post', { id: post.id });
                });
            });

            describe('GET /api/posts/:id', () => {
                it('should get a post by id', async () => {
                    const post = await createPost(prisma, user.id);

                    const response = await client.authGet(`/api/posts/${post.id}`, token);

                    assertResponse(response)
                        .ok()
                        .dataContains({ id: post.id, title: post.title });
                });

                it('should return 404 for non-existent post', async () => {
                    const response = await client.authGet('/api/posts/99999', token);

                    assertResponse(response).notFound();
                });
            });

            describe('PATCH /api/posts/:id', () => {
                it('should update a post', async () => {
                    const post = await createPost(prisma, user.id);

                    const response = await client.authPatch(
                        `/api/posts/${post.id}`,
                        { title: 'Updated Title' },
                        token
                    );

                    assertResponse(response)
                        .ok()
                        .dataContains({ title: 'Updated Title' });
                });

                it('should reject update by non-owner', async () => {
                    const otherUser = await createUser(prisma);
                    const post = await createPost(prisma, otherUser.id);

                    const response = await client.authPatch(
                        `/api/posts/${post.id}`,
                        { title: 'Hacked' },
                        token
                    );

                    assertResponse(response).forbidden();
                });
            });

            describe('DELETE /api/posts/:id', () => {
                it('should delete a post', async () => {
                    const post = await createPost(prisma, user.id);

                    const response = await client.authDelete(`/api/posts/${post.id}`, token);

                    assertResponse(response).noContent();

                    // Verify soft delete
                    const deleted = await prisma.post.findUnique({ where: { id: post.id } });
                    expect(deleted.isDeleted).toBe(true);
                });
            });

            describe('GET /api/posts', () => {
                it('should list posts with pagination', async () => {
                    // Create 25 posts
                    for (let i = 0; i < 25; i++) {
                        await createPost(prisma, user.id);
                    }

                    const response = await client.authGet('/api/posts?page=1&perPage=10', token);

                    assertResponse(response)
                        .ok()
                        .dataLength(10)
                        .paginationEquals({
                            page: 1,
                            perPage: 10,
                            total: 25,
                        });
                });

                it('should filter posts by status', async () => {
                    await createPost(prisma, user.id, { status: 'published' });
                    await createPost(prisma, user.id, { status: 'published' });
                    await createPost(prisma, user.id, { status: 'draft' });

                    const response = await client.authGet('/api/posts?status=published', token);

                    assertResponse(response)
                        .ok()
                        .dataLength(2);
                });

                it('should search posts by title', async () => {
                    await createPost(prisma, user.id, { title: 'JavaScript Guide' });
                    await createPost(prisma, user.id, { title: 'Python Guide' });
                    await createPost(prisma, user.id, { title: 'Other Topic' });

                    const response = await client.authGet('/api/posts?search=Guide', token);

                    assertResponse(response)
                        .ok()
                        .dataLength(2);
                });
            });
        });
    });

    // ============================================================================
    // Webhook Tests
    // ============================================================================

    describe('Webhooks', () => {
        describe('POST /api/webhooks/stripe', () => {
            it('should handle subscription created event', async () => {
                const user = await createUser(prisma, { stripeCustomerId: 'cus_test' });

                const payload = {
                    type: 'customer.subscription.created',
                    data: {
                        object: {
                            id: 'sub_123',
                            customer: 'cus_test',
                            status: 'active',
                            items: {
                                data: [{ price: { id: 'price_pro' } }],
                            },
                        },
                    },
                };

                // In real tests, compute proper Stripe signature
                const response = await client.post('/api/webhooks/stripe', payload, {
                    'Stripe-Signature': 'test_signature',
                });

                assertResponse(response).ok();

                // Verify subscription was created
                await assertExistsInDb(prisma, 'subscription', {
                    userId: user.id,
                    stripeSubscriptionId: 'sub_123',
                });
            });
        });
    });

    // ============================================================================
    // File Upload Tests
    // ============================================================================

    describe('File Uploads', () => {
        let user;
        let token;

        beforeEach(async () => {
            user = await createUser(prisma);
            token = createAuthToken(user.id);
        });

        it('should upload a file', async () => {
            const response = await client
                .authPost('/api/files/upload', {}, token)
                .attach('file', Buffer.from('Hello, World!'), 'test.txt');

            assertResponse(response)
                .created()
                .hasKey('id', 'url', 'filename');
        });

        it('should reject file too large', async () => {
            const largeBuffer = Buffer.alloc(11 * 1024 * 1024); // 11 MB

            const response = await client
                .authPost('/api/files/upload', {}, token)
                .attach('file', largeBuffer, 'large.bin');

            assertResponse(response).status(413);
        });

        it('should reject invalid file type', async () => {
            const response = await client
                .authPost('/api/files/upload', {}, token)
                .attach('file', Buffer.from('malicious'), 'virus.exe');

            assertResponse(response)
                .badRequest()
                .hasError('INVALID_FILE_TYPE');
        });
    });

    // ============================================================================
    // Rate Limiting Tests
    // ============================================================================

    describe('Rate Limiting', () => {
        it('should rate limit excessive requests', async () => {
            const requests = Array(100)
                .fill()
                .map(() => client.get('/api/health'));

            const responses = await Promise.all(requests);
            const rateLimited = responses.filter((r) => r.status === 429);

            expect(rateLimited.length).toBeGreaterThan(0);
        });
    });

    // ============================================================================
    // Concurrent Operations Tests
    // ============================================================================

    describe('Concurrent Operations', () => {
        it('should handle concurrent updates correctly', async () => {
            const user = await createUser(prisma);
            const post = await createPost(prisma, user.id, { viewCount: 0 });
            const token = createAuthToken(user.id);

            // Simulate 10 concurrent view increments
            const requests = Array(10)
                .fill()
                .map(() => client.authPost(`/api/posts/${post.id}/view`, {}, token));

            await Promise.all(requests);

            const updated = await prisma.post.findUnique({ where: { id: post.id } });
            expect(updated.viewCount).toBe(10);
        });
    });
});
